site/en/docs/extensions/mv3/promises/index.md

```diff
@@ -0,0 +1,214 @@
+
+One way to understand how you can use promises in extensions APIs is to compare two equivalent code
+fragments, one using a callback and one using a promise. The following example shows this
+comparison:
+
+<!--
+// --- Standard callback implementation ---
+// --- Promise implementation ---
+-->
+
+#### Standard callback implementation
+
+```js
+
+function openTabOnRight(onComplete) {
+  chrome.tabs.query(queryOptions, function(tabs) {
+    if (chrome.runtime.lastError) {
+      onComplete({error: chrome.runtime.lastError});
+      return;
+    }
+
+    if (!tabs.length) {
+      onComplete();
+      return;
+    };
+
+    chrome.tabs.create({
+      url: 'https://example.com',
+      index: tab[0].index + 1,
+    }, function(tab) {
+      if (chrome.runtime.lastError) {
+        onComplete({error: chrome.runtime.lastError});
+      }
+
+      console.log('tab created', tab);
+      onComplete(tab);
+    });
+  });
+}
+```
+
+#### Promise implementation
+
+```js
+function openTabOnRight() {
+  // Errors are automatically propagated down the promise chain
+  return chrome.tabs.query(queryOptions)
+    .then((tabs) => {
+      if (!tabs.length) return;
+
+      return chrome.tabs.create({
+        url: 'https://example.com',
+        index: tab[0].index + 1,
+      });
+    })
+    .then(tab => {
+      if (!tab) return;
+
+      console.log('tab created', tab);
+      return tab;
+    });
+}
+```
+
+
+<!-- notes here about the code fragments -->
+
+### Error handling
+
+Returning errors works differently depending on if the extension is using a callback or a promise.
+
+#### Error handling with Callbacks
+
+If using a callback, then `chrome.runtime.lastError` is set for the duration of the execution of the
+callback. It is not thrown as a JS Error (which would interrupt JS execution), and is not set
+outside the duration of the callback run (which would result in it being "randomly" set during other
+execution).  The extension would look at the last error like this:
+
+```js
+chrome.tabs.create({...}, (result) => {
+  if (chrome.runtime.lastError) {
+    // Handle last error
+  }
+});
+```
+
+Promises are designed to deliver asynchronous results, both success and failure.  A failure in a
+promise (a promise rejection) is handled differently.  It might look like this:
+
+```js
+chrome.tabs.create({...})
+  .then((result) => {
+    // success case
+  })
+  .catch((e) => {
+    // failure case
+  });
+```
+
+Instead of setting `chrome.runtime.lastError`, we deliver the error directly to the "catch" of the
+promise. We don't set the `lastError` in the success case, because the onFulfilled handler (the
+argument in the `.then())` is never invoked in the failure case.
+
+{% Aside %}
+Extensions APIs don't set `chrome.runtime.lastError()` with promises.
+{% endAside %}
+
+This lets you write async function calls like sync function calls, verifying the failure cases (the
+errors) in the same way.
```

Comment:
I think it may be better to say that "This lets you write async logic in a more synchronous style". 

> …verifying the failure cases (the errors) in the same way.

I don't this this is accurate. Typically errors (e.g. `throw new Error('some description')`) are handled with try/catch. Promises do allow you to _throw_ errors in a synchronous style (throwing inside an async function or a `new Promise(…)` call will route the error throught he promise chain), but _handling_ them requires the use of a error handler in the promise chain.

async/await (specifically async) is the language feature that allows you to use try/catch with promises. That's why above I suggested saying "in a more synchronous style".

Reply:
Modified and simplified this along the suggested lines in 9e0faef.